proxy/src/connection.rs

```diff
@@ -252,24 +262,47 @@ impl Future for Connecting {
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         loop {
-            *self = match self {
-                Connecting::Plaintext { connect, tls } => {
+            self.state = match &mut self.state {
+                ConnectingState::Plaintext { connect, tls } => {
                     let plaintext_stream = try_ready!(connect.poll());
                     set_nodelay_or_warn(&plaintext_stream);
                     match tls.take().expect("Polled after ready") {
                         Conditional::Some(config) => {
-                            let upgrade_to_tls = tls::Connection::connect(
+                            let upgrade = tls::Connection::connect(
                                 plaintext_stream, &config.identity, config.config);
-                            Connecting::UpgradeToTls(upgrade_to_tls)
+                            ConnectingState::UpgradeToTls(upgrade)
                         },
                         Conditional::None(why) => {
                             return Ok(Async::Ready(Connection::plain(plaintext_stream, why)));
                         },
                     }
                 },
-                Connecting::UpgradeToTls(upgrading) => {
-                    let tls_stream = try_ready!(upgrading.poll());
-                    return Ok(Async::Ready(Connection::tls(BoxedIo::new(tls_stream))));
+                ConnectingState::UpgradeToTls(upgrade) => {
+                    match upgrade.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(tls_stream)) => {
+                            let conn = Connection::tls(BoxedIo::new(tls_stream));
+                            return Ok(Async::Ready(conn));
+                        },
+                        Err(e) => {
+                            warn!(
+                                "TLS handshake with {:?} failed: {}\
+                                    -> falling back to plaintext",
+                                self.addr, e,
+                            );
+                            // XXX: We can't get the old connection back
+                            // from the failed upgrade future, so we have
+                            // to try reconnecting.
```

Comment:
I don't really follow this comment. If we've tried to send a TLS handshake on a connection that refused it, there's nothing else that can be done with that connection.

Reply:
Okay, I wasn't sure if the socket could be reused. I'll remove this comment.